src/data/rel.lean

```diff
@@ -2,16 +2,36 @@
 Copyright (c) 2018 Jeremy Avigad. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Jeremy Avigad
-
-Operations on set-valued functions, aka partial multifunctions, aka relations.
 -/
 import data.set.lattice
 
-variables {α : Type*} {β : Type*} {γ : Type*}
+/-!
+# Relations
+
+This file defines bundled relations. A relation between `α` and `β` is a function `α → β → Prop`.
+Relations are also known as set-valued functions, or partial multifunctions.
+
+## Main declarations
+
+* `rel α β`: Relation between `α` and `β`.
+* `rel.inv`: `r.inv` is the `rel β α` obtained by swapping the arguments of `r`.
+* `rel.dom`: Domain. `x ∈ r.dom` iff there exists `y` such that `r x y`.
+* `rel.codom`: Codomain, or range. `y ∈ r.codom` iff there exists `x` such that `r x y`.
+* `rel.comp`: Relation composition. Note that the arguments order follows the `category_theory/`
+  one, so `r.comp s x z ↔ ∃ y, r x y ∧ s y z`.
+* `rel.image`: Image of a set under a relation.
+* `rel.preimage`: Preimage of a set under a relation. Note that `r.preimage = r.inv.image`.
+* `rel.core`: Core of a set. For `s : set β`, `r.core s` is the set of `x : α` such that all `y`
+  related to `x` are in `s`.
+* `rel.restrict_domain`: Domain-restriction of a relation.
```

Comment:
I think a tiny bit more detail here could be nice, currently it could mean either restricting the relation to its domain, or restricting the domain to any subtype